### PR TITLE
Improve ordering of coarse cells in parallel::distributed::Triangulation

### PR DESCRIPTION
**Pull request by kronbichler**
On parallel computations where the coarse mesh consists of many cells (say a few thousands or more), the numbering of coarse mesh cells is based on the Cuthill-McKee algorithm which creates very thin layers. If the number of processors is of similar magnitude as the number of layers, which is quite frequent especially in 3D, you get horrible surface-to-volume ratios.

Therefore, I designed a better renumbering algorithm that aims to reproduce what p4est does inside each tree, namely hierarchical numbering. To this end, I collect nodes and its direct neighbors and do this recursively, which gives me a similar approach. Even though the partitions are still worse than what you would get with METIS or Zoltan with the given number of partitions, the numbering is flexible with respect to the rebalancing that happens when mesh adaptation is done. (At least, I am not aware of any good partitioning tool towards that goal.)

To appreciate this, check these two figures (old versus new), run on 17 processors:
old
new

And to add more support to my claim and where I actually observed that something improved was necessary, run times on medium sized computations with 256-2048 cores and 8800 coarse mesh cells (the above figure but used in 3D with GridGenerator::extrude_triangulation) improved by 30-50 percent (!). Partly because ILU's are getting better, partly because communication is much improved.

### CODE REVIEW

**Comment by kronbichler** [comment]
One thing I should add: In order to improve the quality of the ordering, I use a new function GridTools::get_vertex_connectivity_of_cells instead of connectivity over faces only - this way, a subdivided hypercube mesh actually ends up begin numbered quite close to the hierarchical numbering when starting from one cell only and refining...

File: include/deal.II/grid/grid_tools.h [hunk 1]
```diff
@@ -573,9 +573,32 @@ namespace GridTools
...
+
+  /**
+   * As above, but filling a SparsityPattern object instead.
+   *
+   * @deprecated
+   */
+  template <int dim, int spacedim>
+  void
+  get_face_connectivity_of_cells (const Triangulation<dim, spacedim> &triangulation,
                                   SparsityPattern                    &connectivity);
```
Comment (Rombur) [inline include/deal.II/grid/grid_tools.h]: You forgot to add  DEAL_II_DEPRECATED

File: tests/distributed_grids/2d_coarse_grid_01.debug.output [hunk 2]
```diff
@@ -64,11 +64,11 @@ DEAL:2d::Checksum: 3932161
            -9.09745693e-01  -8.64258409e-01   0.00000000e+00
            -2.09025431e+00   1.98574162e+00   0.00000000e+00
            -9.09745693e-01   8.64258409e-01   0.00000000e+00
-            2.09025431e+00  -1.98574162e+00   0.00000000e+00
-            2.09025431e+00   1.98574162e+00   0.00000000e+00
-            9.09745693e-01  -8.64258409e-01   0.00000000e+00
-            9.09745693e-01   8.64258409e-01   0.00000000e+00
            -8.34745646e-01  -8.34745646e-01   0.00000000e+00
+            8.34745646e-01  -8.34745646e-01   0.00000000e+00
```
Comment (tjhei) [inline tests/distributed_grids/2d_coarse_grid_01.debug.output]: Why do the coordinates of the cells change here and not just the ordering?

**Comment by tjhei** [comment]
Very cool, I have been wondering about this in the past. :+1

**Comment by Rombur** [comment]
This is very nice. I like it a lot !!

File: tests/distributed_grids/2d_coarse_grid_01.debug.output (hunk 2 above)
Comment (kronbichler) [inline tests/distributed_grids/2d_coarse_grid_01.debug.output]: These tests just output the first 20 lines (see tests/distributed/coarse_grid_common.h, line 54), and so there are different points/cells that get into the file.

File: include/deal.II/grid/grid_tools.h (hunk 1 above)
Comment (kronbichler) [inline include/deal.II/grid/grid_tools.h]: Fixed, thanks.

**Comment by bangerth** [comment]
Very cool -- I like this a lot!

**Pull request merged**